File: backend/bc_expense_tax.py
"""BC net-of-GST / ITC helpers — mirrors frontend bc-expense-tax.ts for SQL exports."""

from __future__ import annotations

from models import Expense, Jurisdiction

EPS = 1e-6


def _round_money(n: float) -> float:
    return round(n, 2)


def should_bypass_canadian_itc_estimation(e: Expense) -> bool:
    """USD-native or USA jurisdiction: never infer Canadian GST/HST ITC."""
    cur = (e.original_currency or "CAD").upper()
    if cur == "USD":
        return True
    if e.jurisdiction == Jurisdiction.USA:
        return True
    return False


def is_usd_payment(e: Expense) -> bool:
    """Original currency USD — no Canadian GST/HST/PST in ITC, net expense, or export tax columns."""
    return (e.original_currency or "CAD").upper() == "USD"


def apply_usd_payment_tax_policy(expense: Expense) -> None:
    """Mutate expense: USD payments cannot store Canadian tax (all categories)."""
    if not is_usd_payment(expense):
        return
    expense.gst_amount = 0.0
    expense.hst_amount = 0.0
    expense.pst_amount = 0.0
    expense.tax_amount = 0.0
    expense.gst_itc_estimated = False


def expense_has_manual_or_parsed_tax(e: Expense) -> bool:
    gst = e.gst_amount or 0
    hst = e.hst_amount or 0
    pst = e.pst_amount or 0
    tax = e.tax_amount or 0
    return gst > EPS or hst > EPS or pst > EPS or tax > EPS


def get_effective_gst_hst_for_itc(e: Expense) -> tuple[float, float]:
    gst_raw = e.gst_amount or 0
    hst_raw = e.hst_amount or 0
    tax_amt = e.tax_amount or 0
    if gst_raw <= EPS and hst_raw <= EPS and tax_amt > EPS:
        return 0.0, tax_amt
    return gst_raw, hst_raw


def effective_recoverable_itc_cad(e: Expense) -> float:
    if is_usd_payment(e):
        return 0.0
    if getattr(e, "gst_itc_estimated", None) is True:
        return 0.0
    if not expense_has_manual_or_parsed_tax(e):
        return 0.0
    g, h = get_effective_gst_hst_for_itc(e)
    return _round_money(g + h)


def net_expense_cad(e: Expense) -> float:
    gross = e.cad_amount or 0
    itc = effective_recoverable_itc_cad(e)
    return _round_money(max(0.0, gross - itc))


def recorded_canadian_tax_parts(e: Expense) -> tuple[float, float, float]:
    """GST, effective HST (with legacy tax_amount), PST. USD / legacy system estimate → zero."""
    if is_usd_payment(e):
        return 0.0, 0.0, 0.0
    if getattr(e, "gst_itc_estimated", None) is True:
        return 0.0, 0.0, 0.0
    gst = e.gst_amount or 0
    hst_raw = e.hst_amount or 0
    tax_amt = e.tax_amount or 0
    if gst <= EPS and hst_raw <= EPS and tax_amt > EPS:
        return gst, tax_amt, e.pst_amount or 0
    return gst, hst_raw, e.pst_amount or 0


def itc_source_label(_e: Expense) -> str:
    return "Manual / Receipt"

File: frontend/src/lib/bc-expense-tax.ts
/**
 * Canadian ITC: only from receipt/OCR or manually entered GST/HST (and legacy tax_amount → HST).
 * No category- or gross-based estimation.
 */

const EPS = 1e-6;

/** Minimal expense shape for tax helpers (avoids circular imports with firebase-api). */
export type ExpenseTaxFields = {
  cad_amount?: number | null;
  category?: string;
  jurisdiction?: string;
  original_currency?: string;
  currency?: string;
  gst_amount?: number;
  hst_amount?: number;
  pst_amount?: number;
  tax_amount?: number;
};

export function roundMoney(n: number): number {
  return Math.round(n * 100) / 100;
}

/** True if any tax field was populated (manual entry, OCR, or legacy total tax). */
export function expenseHasManualOrParsedTax(e: Pick<ExpenseTaxFields, "gst_amount" | "hst_amount" | "pst_amount" | "tax_amount">): boolean {
  const gst = e.gst_amount ?? 0;
  const hst = e.hst_amount ?? 0;
  const pst = e.pst_amount ?? 0;
  const tax = e.tax_amount ?? 0;
  return (
    gst > EPS || hst > EPS || pst > EPS || tax > EPS
  );
}

/**
 * True when Canadian GST/HST/PST must never be inferred (US spend / USD native).
 * CAD amount may exist from FX — ITC still zero for CRA.
 */
export function shouldBypassCanadianItcEstimation(
  originalCurrency: string | undefined,
  jurisdiction: string | undefined,
): boolean {
  const cur = (originalCurrency || "CAD").toUpperCase();
  const jur = (jurisdiction || "").toLowerCase();
  if (cur === "USD") return true;
  if (jur === "usa" || jur === "us") return true;
  return false;
}

/** True when the expense was paid in USD — no Canadian GST/HST/PST in calculations or exports. */
export function expenseIsUsdPayment(
  e: Pick<ExpenseTaxFields, "original_currency" | "currency">,
): boolean {
  return (e.original_currency || e.currency || "CAD").toUpperCase() === "USD";
}

/** GST/HST/PST as stored for UI and summaries — USD rows are always zero. */
export function storedCanadianTaxForDisplay(
  e: ExpenseItcMeta,
): { gst: number; hst: number; pst: number } {
  if (expenseIsUsdPayment(e)) return { gst: 0, hst: 0, pst: 0 };
  if (e.gst_itc_estimated === true) return { gst: 0, hst: 0, pst: 0 };
  const gst = e.gst_amount ?? 0;
  const hstRaw = e.hst_amount ?? 0;
  const taxAmt = e.tax_amount ?? 0;
  const hst = gst <= EPS && hstRaw <= EPS && taxAmt > EPS ? taxAmt : hstRaw;
  return { gst, hst, pst: e.pst_amount ?? 0 };
}

/** Persisted fields to clear for USD payments (all categories). */
export function usdPaymentZeroTaxWritePatch(): {
  gst_amount: number;
  hst_amount: number;
  pst_amount: number;
  tax_amount: number;
  gst_itc_estimated: false;
} {
  return {
    gst_amount: 0,
    hst_amount: 0,
    pst_amount: 0,
    tax_amount: 0,
    gst_itc_estimated: false,
  };
}

export type ExpenseItcMeta = ExpenseTaxFields & { gst_itc_estimated?: boolean };

const zeroTaxClearPatch = (): Record<string, unknown> => ({
  ...usdPaymentZeroTaxWritePatch(),
});

/**
 * Category bulk-edit: never apply ITC estimates. Strip legacy system-estimated tax; USD policy unchanged.
 */
export function mergeBcItcForCategoryChange(
  existing: ExpenseItcMeta,
  newCategory: string,
): Record<string, unknown> {
  const out: Record<string, unknown> = { category: newCategory };
  if (expenseIsUsdPayment(existing)) {
    Object.assign(out, usdPaymentZeroTaxWritePatch());
    return out;
  }
  const cur = existing.original_currency || existing.currency;
  if (shouldBypassCanadianItcEstimation(cur, existing.jurisdiction)) {
    if (existing.gst_itc_estimated === true) Object.assign(out, zeroTaxClearPatch());
    return out;
  }
  if (existing.gst_itc_estimated === true) {
    Object.assign(out, zeroTaxClearPatch());
    return out;
  }
  return out;
}

/** GST+HST stored or inferred for legacy rows (tax_amount → HST). */
export function getEffectiveGstHstForItc(
  e: Pick<ExpenseTaxFields, "gst_amount" | "hst_amount" | "tax_amount">,
): { gst: number; hst: number } {
  const gstRaw = e.gst_amount ?? 0;
  const hstRaw = e.hst_amount ?? 0;
  const taxAmt = e.tax_amount ?? 0;
  if (gstRaw <= EPS && hstRaw <= EPS && taxAmt > EPS) {
    return { gst: 0, hst: taxAmt };
  }
  return { gst: gstRaw, hst: hstRaw };
}

/** Recoverable ITC (GST + HST) only from entered/parsed tax lines — never from category or gross. */
export function getEffectiveRecoverableItcCad(e: ExpenseItcMeta): number {
  if (expenseIsUsdPayment(e)) {
    return 0;
  }
  if (e.gst_itc_estimated === true) {
    return 0;
  }
  if (!expenseHasManualOrParsedTax(e)) {
    return 0;
  }
  const { gst, hst } = getEffectiveGstHstForItc(e);
  return roundMoney(gst + hst);
}

/** Net operating expense for P&L display: gross CAD − recoverable ITC (PST stays in gross). */
export function getNetExpenseCad(e: ExpenseItcMeta): number {
  const gross = e.cad_amount ?? 0;
  const itc = getEffectiveRecoverableItcCad(e);
  return roundMoney(Math.max(0, gross - itc));
}

/** Export / audit label — category-based estimation is disabled. */
export function getItcSourceLabel(_e: ExpenseItcMeta): "Auto-Estimated" | "Manual / Receipt" {
  return "Manual / Receipt";
}
